### ISSUE TITLE
Edit 'Sampled Flag' section to be more concise.

### ISSUE THREAD

**Issue by danielkhan** [w3c/trace-context#303]
Right now, the section covering the `sampled` flag is as follows:

```
#### Sampled Flag (00000001)

When set, the least significant bit documents that the caller may have recorded
trace data. A caller who does not record trace data out-of-band leaves this flag
unset.

Many distributed tracing scenarios may be broken when only a subset of calls
participated in a <a>distributed trace</a> were recorded. At certain load
recording information about every incoming and outgoing request becomes
prohibitively expensive. Making a random or component-specific decision for data
collection will lead to fragmented data in every <a>distributed trace</a>. Thus
it is typical for tracing vendors and platforms to pass recording decision for
given distributed trace or information needed to make this decision.

There is no consensus on what is the best algorithm to make a recording
decision. Various techniques include: probability sampling (sample 1 out of 100
<a>distributed traces</a> by flipping a coin), delayed decision (make collection
decision based on duration or a result of a call), deferred sampling (let callee
decide whether information about this request need to be collected). There are
variations and customizations of every technique which can be tracing vendor
specific or application defined.

Field `tracestate` is designed to handle the variety of techniques for making
recording decision specific (along any other specific information) for a given
tracing system or a platform. Flag `sampled` is introduced for better
interoperability between vendors. It allows to communicate recording decision
and enable better experience for the customer.

For example, when SaaS services participate in <a>distributed trace</a> - this
service has no knowledge of tracing system used by its caller. But this service
may produce records of incoming requests for monitoring or troubleshooting
purposes. Flag `sampled` can be used to ensure that information about requests
that were marked for recording by caller will also be recorded by SaaS service.
So caller can troubleshoot the behavior of every recorded request.

Flag `sampled` has no restriction on its mutations except that it can only be
mutated when `parent-id` was updated. See section "Mutating the traceparent
field". However there are set of suggestions that will increase vendors
interoperability.

1. If component made definitive recording decision - this decision SHOULD be
   reflected in `sampled` flag.
2. If component needs to make a recording decision - it SHOULD respect
   `sampled` flag value. Security considerations should be applied to protect
   from abusive or malicious use of this flag - see security section.
3. If component deferred or delayed decision and only a subset of telemetry will
   be recorded - flag `sampled` should be propagated unchanged. And set to `0`
   as a default option when trace is initiated by this component. There are two
   additional options:
    1. Component that makes deferred or delayed recording decision may
       communicate priority of recording by setting `sampled` flag to `1` for a
       subset of requests.
    2. Component may also fall back to probability sampling to set flag
       `sampled` to `1` for the subset of requests.
```
On various occasions, for example in #265 we came to the conclusion, that this text contains too much reasoning about why this flag exists and how to use it.

I'd like to propose to trim down this section.

My assumptions are
1. We should not need to explain why sampling is used as implementors won't base their data collection algorithms on our spec.
2. We should not need to explain different sampling strategies, as implementors that need sampling and rely on that flag should be fully aware of this.
3. We should not need to explain use-cases (like interop) as implementors should be able to derive them if the flag has a clear semantic.

tl;dr: Those that need that flag to propagate sampling decisions should know how to use it. Those that don't need that flag should know which information it contains, how they might incorporate it and how to propagate it.

That's why I'd like to the discuss the following - admitetly drastic - change:

```
#### Sampled Flag (00000001)

Sampling is a technique to reduce overhead and the amount of collected data during
tracing.
A span is called 'sampled' when it was recorded by a tracing system.

The `sampled` flag provides a way to propagate such sampling decisions.

##### Rules

- The `sampled` flag MUST only be mutated when `parent-id` is updated as well.
- If a span was sampled, the caller SHOULD set the least significant bit to 1.
- If a span was not sampled, the caller SHOULD set the least significant bit to 0.
- If a caller deferred or delayed the sampling decision, the `sampled` flag
  SHOULD be propagated unchanged or respectively SHOULD be set to `0` when the
  trace is initiated by this caller.
- If a callee needs to make a sampling decision, it SHOULD respect this flag.

  Security considerations should be applied to protect from abusive or malicious
  use of this flag - see [security considerations](#security-considerations).
``` 

I didn't create a PR because I wanted to open this up for a broader discussion before we incorporate any changes.

FYI @SergeyKanzhelev, @yurishkuro, @adriancole (especially because you mentioned, that the semantic of the flag has changed along the way and I want to see if we are back to the semantic the industry uses)

**Comment by yurishkuro** [comment]
I agree text needs to be simplified. Rationalizations can exist in separate ADR documents.

In terms of semantic meaning, is it fair to say that we always treat the caller as untrusted? Many existing tracers (Jaeger, and afaik zipkin-brave, OpenCensus) normally only operate in trusted mode, ie always respect the caller's sampling decision. Does it mean they always must use tracestate to record "trusted" sampling flag?

**Comment by SergeyKanzhelev** [comment]
> I agree text needs to be simplified. Rationalizations can exist in separate ADR documents.

I agree in principle, but sampling concept and the importance of propagating the flag is not a concept that people outside of tracing world understand very well. Since we intend this spec for the wider community - I believe it is important to have some introduction into the problem this flag is solving.

Proposed text fails to explain why the flag should ever be propagated. How it can be consumed by the system you operate (like SaaS platform).

Also, the proposed text introduces the word `span` that we intentionally wanted to avoid and even renamed `span-id` to `parent-id`.

**Comment by yurishkuro** [comment]
I am not convinced that the existing text explains the concept of sampling any better to people who don't know what it is. Sampling is a complicated topic, should the spec be really responsible for educating people about it? Why not refer people to something where it is explained more in-depth, such as http://www.pdl.cmu.edu/PDL-FTP/SelfStar/CMU-PDL-14-102.pdf ?

**Comment by danielkhan** [comment]
@SergeyKanzhelev - thank you for pointing out the spans. I didn't know that. How would you translate 'a span' - just looking for the right nomenclature here.

I agree with @yurishkuro that it may be out of scope to educate people within the spec itself. Having a little longer intro and then linking to a scientific document or adding the explanation to rationale sounds like a good compromise to me.

**Comment by danielkhan** [comment]
> In terms of semantic meaning, is it fair to say that we always treat the caller as untrusted? Many existing tracers (Jaeger, and afaik zipkin-brave, OpenCensus) normally only operate in trusted mode, ie always respect the caller's sampling decision. Does it mean they always must use tracestate to record "trusted" sampling flag?

As far as I understand the discussed semantic so far, it is up to the tracing system to determine if a caller is trusted or not. @SergeyKanzhelev - what's your opinion on that?

**Comment by SergeyKanzhelev** [comment]
@danielkhan from my perspective spec needs to educate people on a topic of sampling. At least enough to explain why you need to propagate sampling decision for head-based sampling, why it's not always working and what may be the risk associating with trusting this sampling decision. It may be a separate section, but without it, the spec is hard to understand. If there is a good specification explaining sampling - it will definitely be beneficial to refer to it. My understanding of the links from the specs - it must link to the more mature specs or something that will not be changed. Perhaps specs from other standardization bodies. External links to blogs or Wikipedia pages are hard (if not impossible) to make normative. 

As for trusted or not trusted - my read of a current spec is that it presumes trust, but talks about risks.

For `span` - you can just say "information about incoming and outgoing requests". The term request is simpler to understand in the context of W3C. The argument against `span` was events and logs based systems that don't have this concept.

**Comment by SergeyKanzhelev** [comment]
I missed link from Yuri. Looks good enough to refer to as it seems to be some publication. I haven't read it yet though

**Comment by SergeyKanzhelev** [comment]
> I am not convinced that the existing text explains the concept of sampling any better to people who don't know what it is. Sampling is a complicated topic, should the spec be really responsible for educating people about it?

I think the message spec needs to deliver to explain the need for the flag is:

1. It may be expensive to record information about all requests
2. It is important to preserve as much as possible end-to-end distributed traces
3. Typical use - head based sampling. Behavior - if you see if `1` - record, see `0` - don't
4. There are other sampling strategies. For sampling. Flag behavior largely undefined for those cases
5. There are risks associated with trusting this flag (security section).

We use this spec to communicate requirements to different services and understanding of those bullet points is very important. People who interested to learn more will follow the link.

**Comment by danielkhan** [comment]
> I think the message spec needs to deliver to explain the need for the flag is:
> 
> 1. It may be expensive to record information about all requests

Can you think of a persona that does not know this and needs it when implementing the spec?

> 2. It is important to preserve as much as possible end-to-end distributed traces

For whom? This is an implementation detail of a monitoring system and it's up to them to preserve as much traces for their usecase.

> 3. Typical use - head based sampling. Behavior - if you see if `1` - record, see `0` - don't

This is in the text.

> 4. There are other sampling strategies. For sampling. Flag behavior largely undefined for those cases

Then why mention it in a spec?

> 5. There are risks associated with trusting this flag (security section).

This is in the text.

> We use this spec to communicate requirements to different services and understanding of those bullet points is very important. People who interested to learn more will follow the link.

I think that we maybe have different personas reading the spec in mind and I really want to discuss that because I am not sure what's right. Also maybe broader @nikmd23, @mtwo.

tl;dr: To me, sampling is a typical case of "If you don't know what it is, you obviously don't sample and you should most probably just propagate it according to the spec."

**Comment by SergeyKanzhelev** [comment]
@danielkhan I'm talking from experience of talking with service owners who are not thinking about tracing every day as ourselves =). Even the concept that "logs" may be too large and expensive sometimes is a new concept. But they are still responsible for exposing telemetry and propagating context.

**Comment by yurishkuro** [comment]
I think service owners are very unlikely users of this spec, at best they will use a library like OpenTelemetry that will handle the context for them.

**Comment by SergeyKanzhelev** [comment]
I mean SaaS services like Azure services. Same may apply for boxed products like meshes.

**Comment by AloisReitbauer** [comment]
We should separate the definition of the 'protocol' in the spec and a wider explanation of core concepts. The explanation makes sense but should go in an implementors guide. The spec is to define WHAT and HOW things are done and not WHY. 

If we do this everywhere the spec gets too verbose and will confuse people more than it helps. Let us keep things simple.

**Comment by SergeyKanzhelev** [comment]
@AloisReitbauer I agree with you. When you start writing it as lean as possible - you end up with the few paragraphs explaining the concept and basic expectation from this bit. See the list I posted. You want to simply say sampling bit helps to reduce overhead. So you need to explain what the word "sampling" mean as it's not a common knowledge. So you need to explain the concept of data recording and how expensive it may be. You also want to discourage people who want to keep a local counter and sample 1 out of 100 - you need to explain why. Then you tell about basic strategy of trusting your parent. Finally you add couple words that there are other sampling strategies and some security implications. Thus trusting your parent is SHOULD, not MUST. I believe what @danielkhan is almost good as it covers most of it, but need a bit more wording defining concepts.

Again, all these clarifications I ask for coming from real conversations with real people. 

@nikmd23 can also chime in here.

**Comment by nikmd23** [comment]
My sense is that there's probably a happy medium to be achieved here. The current text is too verbose, and the proposed text a bit terse. I like the idea of linking to a resource with further explanation, but we've run into difficulties with that in the past. The W3C would prefer that we inline that explanation into our document. 

Perhaps the right move here is to go with the proposed text, but also update the spec's abstract where we define the problem space. The abstract could specifically mention sampling, since it's almost a secondary problem. Perhaps something like this?

```
This specification defines standard headers and value 
format to propagate context information that enables 
distributed tracing scenarios. The specification standardizes 
how context information is sent and modified between 
services. Context information uniquely identifies individual 
requests in a distributed system and also defines a means 
to add and propagate provider-specific context information. 

In addition, this specification defines a mechanism to signal 
sampling behavior to downstream services so those services 
may choose to opt out of recording trace information in 
situations that would be expensive, while having the least 
impact to the usefulness of the overall trace as possible.
```

Thoughts on this proposal? (My text is draft text, I'm looking for 1) feedback on the approach, and 2) on the text itself - we can clean that up in a PR.)

**Comment by nikmd23** [comment]
Actually, better yet - why don't we create a "Sampling Consideration" section like we have for privacy and security?

**Comment by SergeyKanzhelev** [comment]
@nikmd23 I'm fine moving explanation of concepts to other section. Do you agree in principle that spec needs to explain what sampling means and how propagation of a sampling decision makes a better traces?

**Comment by nikmd23** [comment]
I don't see how it could hurt as long as it isn't interrupting the flow of the how/what.

**Comment by mtwo** [comment]
I prefer the new terse text. I'm indifferent about mentioning sampling in the abstract ... we should do this if it means that it settles everyone's concerns.

**Comment by SergeyKanzhelev** [comment]
@mtwo which version of terse text? there are at least two proposals in this thread =). Ultimately, we all agree for shorter text. The question is how many concepts needs explanation and in which sections those explanation will reside

**Comment by AloisReitbauer** [comment]
I want to point out again that we need to separate two topics. The goal of a spec is NOT to educate people at the conceptual level. If people don't understand the flag, they better leave it as it is. 

The spec should define HOW it is used and which situations are supported. This spec is for implementors. I assume that they have an understanding of the underlying technology. 

The main problem with the sampling flag is that we watered it's meaning down to a point where it is hard to understand what it exactly means. 

I propose we make the text short and concise to something like  ```The sampling flag is used to communicate a sampling decision to downstream services. If the flag is set on an incoming request, the transaction SHOULD be considered to be sampled. If the tracer explicitly decides to not sample a request it SHOULD set the flag to 0```

If we really think that we need to educate people because things are too complicated than most likely they are too complicated and it is our job to make them easier and simpler to understand.

**Comment by AloisReitbauer** [comment]
So are we agreeing on Daniel's text now?

**Comment by mtwo** [comment]
I don't have strong opinions about what this section should look like ... @SergeyKanzhelev ?

**Comment by danielkhan** [comment]
Obsolete
